modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java

```diff
@@ -627,22 +720,104 @@ public void groupKey(int grpId, byte[] encGrpKey) {
         return withMasterKeyChangeReadLock(() -> getSpi().getMasterKeyName());
     }
 
+    /** {@inheritDoc} */
+    @Override public IgniteFuture<Void> changeCacheGroupKey(Collection<String> cacheOrGrpNames) {
+        A.notEmpty(cacheOrGrpNames, "cacheOrGrpNames");
+
+        synchronized (opsMux) {
+            if (stopped) {
+                return new IgniteFinishedFutureImpl<>(new IgniteException("Cache group key change was rejected. " +
+                    "Node is stopping."));
+            }
+
+            return grpKeyChangeProc.start(cacheOrGrpNames);
+        }
+    }
+
+    /**
+     * @param grpIds Cache group IDs.
+     * @param keyIds Encryption key IDs.
+     * @param keys Encryption keys.
+     * @throws IgniteCheckedException If failed.
+     */
+    protected void changeCacheGroupKeyLocal(int[] grpIds, byte[] keyIds, byte[][] keys) throws IgniteCheckedException {
+        Map<Integer, Byte> encryptionStatus = U.newHashMap(grpIds.length);
+
+        for (int i = 0; i < grpIds.length; i++)
+            encryptionStatus.put(grpIds[i], keyIds[i]);
+
+        WALPointer ptr = ctx.cache().context().wal().log(new ReencryptionStartRecord(encryptionStatus));
+
+        if (ptr != null)
+            ctx.cache().context().wal().flush(ptr, false);
+
+        for (int i = 0; i < grpIds.length; i++) {
+            int grpId = grpIds[i];
+
+            CacheGroupContext grp = ctx.cache().cacheGroup(grpId);
+
+            if (grp == null)
+                continue;
+
+            int newKeyId = keyIds[i] & 0xff;
+
+            synchronized (metaStorageMux) {
+                // Set new key as key for writing.
+                GroupKey prevGrpKey = grpKeys.changeActiveKey(grpId, newKeyId);
+
+                writeGroupKeysToMetaStore(grpId);
+
+                if (ptr != null) {
+                    grpKeys.reserveWalKey(grpId, prevGrpKey.unsignedId(), ctx.cache().context().wal().currentSegment());
+
+                    writeTrackedWalIdxsToMetaStore();
+                }
+            }
+
+            reencryptGroups.put(grpId, pageScanner.pagesCount(grp));
+
+            if (log.isInfoEnabled())
+                log.info("New encryption key for group was added [grpId=" + grpId + ", keyId=" + newKeyId + "]");
+        }
+
+        startReencryption(encryptionStatus.keySet());
+    }
+
+    /**
+     * @param grpId Cache group ID.
+     * @return Future that will be completed when reencryption of the specified group is finished.
+     */
+    public IgniteInternalFuture<Void> reencryptionFuture(int grpId) {
+        return pageScanner.statusFuture(grpId);
+    }
+
+    /**
+     * @param grpId Cache group ID.
+     * @return {@code True} If the specified cache group should be re-encrypted.
+     */
+    public boolean reencryptionRequired(int grpId) {
```

Comment:
All usages of this method want to check that reencryption in progress. Let's rename it to something like `reencryptionInProgress`